### ISSUE TITLE
Error converting access token to token with Service Principal

### ISSUE THREAD

**Issue by jdfindwise** [terraform-providers/terraform-provider-azurerm#474]
### Terraform Version
Terraform v0.10.8

### Affected Resource(s)
Login issue I believe.

### Debug Output
Please provider a link to a GitHub Gist containing the complete debug output: https://gist.github.com/jdfindwise/6f1122400bd311b98331400b65f1863b

### Expected Behavior
I login with Service Principal using the command `az login --service-principal -u <user> --password <password> --tenant <tenant>`

Login was successful.

I expected the `terraform plan` to show me the plan, just as when I login using `az login` with my personal account.

### Actual Behavior
Error running plan, as showed in the provided gist.

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. Login using a service principal and `az login`
2. `terraform plan`

**Comment by tombuildsstuff** [comment]
Hey @jdfindwise 

Thanks for opening this issue :)

So that we can investigate this further - would it be possible to paste the output from `az --version` and tell us which OS/platform you're on?

Thanks!

**Comment by jdfindwise** [comment]
This is actually running on Ubuntu in Windows, but I installed Linux version of both Terraform and AZ CLI 2.0

```
azure-cli (2.0.20)

acr (2.0.14)
acs (2.0.18)
appservice (0.1.19)
backup (1.0.2)
batch (3.1.6)
batchai (0.1.2)
billing (0.1.6)
cdn (0.0.10)
cloud (2.0.9)
cognitiveservices (0.1.9)
command-modules-nspkg (2.0.1)
component (2.0.8)
configure (2.0.12)
consumption (0.1.6)
container (0.1.12)
core (2.0.20)
cosmosdb (0.1.14)
dla (0.0.13)
dls (0.0.16)
eventgrid (0.1.5)
extension (0.0.5)
feedback (2.0.6)
find (0.2.7)
interactive (0.3.11)
iot (0.1.13)
keyvault (2.0.13)
lab (0.0.12)
monitor (0.0.11)
network (2.0.17)
nspkg (3.0.1)
profile (2.0.15)
rdbms (0.0.8)
redis (0.2.10)
resource (2.0.17)
role (2.0.14)
servicefabric (0.0.5)
sql (2.0.14)
storage (2.0.18)
vm (2.0.17)

Python location '/opt/az/bin/python3'
Extensions directory '/home/dev/.azure/cliextensions'

Python (Linux) 3.6.1 (default, Oct 18 2017, 20:41:18)
[GCC 4.8.4]
```

**Comment by tombuildsstuff** [comment]
@jdfindwise thanks for that - out of interest do you also see this problem natively on Windows, or is it limited to the Windows Subsystem for Linux/Bash on Windows?

**Comment by jdfindwise** [comment]
I just installed AZ CLI 2.0 and Terraform v0.10.8 with azurerm (0.3.2) natively on Windows, same issue.

**Comment by adampats** [comment]
For anyone that comes across this issue in the meantime - the workaround I have found is to re-generate the application Service Principle using a separate `az login` (in my case I use my "root" account) and then rerun my TF commands.

```sh
az ad sp create-for-rbac --role="Contributor" --scopes="/subscriptions/00000000-0000-0000-0000-000000000000"
```

Then I source the necessary secret values from a local script:

```sh
export CLIENT_ID='00000000-0000-0000-0000-000000000000'
export CLIENT_SECRET='00000000-0000-0000-0000-000000000000'
export TENANT_ID='00000000-0000-0000-0000-000000000000'
```
And remove the temporary SP when I'm done:

```sh
az ad sp delete --id 00000000-0000-0000-0000-000000000000
```
FWIW, I am running Azure CLI 2.0 with TF 0.10.8 on OS X 10.13.

**Comment by hantuzun** [comment]
I face this issue with the latest Terraform version 0.11.2 on Centos as well: https://releases.hashicorp.com/terraform/0.11.2/terraform_0.11.2_linux_amd64.zip

**Comment by pmarques** [comment]
Also facing this bug on terraform 011.2 + azure provider 1.0.0 and 1.0.1

**Comment by pmarques** [comment]
@tombuildsstuff I belive this is because we are trying to read tokens with this function, [ToADALToken() : autorest/azure/cli/token.go#L44](https://github.com/Azure/go-autorest/blob/master/autorest/azure/cli/token.go#L44) which doesn't understand Service Principals.

I can use the solution provided by @adampats because I don't have permissions to create Service Principals :(

**Comment by anoff** [comment]
Running on `Terraform v0.11.3`, `provider.azurerm v1.2.0` and `azure-cli (2.0.28)` with the same error 😞

@adampats I am trying to reproduce your solution. Can you elaborate on what exactly you did? Clean environment and just set the ENV vars before running terraform apply?

**Comment by tombuildsstuff** [comment]
@anoff in relation to a comment I saw on another issue: out of interest is your home directory located on a Network Share/are you logged in using Active Directory by any chance?

**Comment by anoff** [comment]
Home directory is local on my MacOS machine, it has some symlinks however if that might also be an issue. My account is managed via Azure AD, yes.

**Comment by jdfindwise** [comment]
My home folder is also an arbitrary folder on my local system, but the Azure account is managed via Azure AD as well.
Could this be specifically related to accounts managed through AD?

**Comment by tombuildsstuff** [comment]
> it has some symlinks however if that might also be an issue

@anoff I'm not sure how it could, but I'd be interested to see an `ls -la` and a `pwd` of `~/.azure` if you can share it

> Could this be specifically related to accounts managed through AD?

@jdfindwise so all Azure accounts are connected to Azure Active Directory (even if that's not connected to an on-prem AD) - as such I don't believe so. My theory was that the Azure CLI parser wasn't loading from networked attached home directory, but I may be wrong (we load from `~/.azure/[file]` - but that should be fine AFAIK)

Thanks!

**Comment by ddarwent** [comment]
Having the same issue on a Windows Build agent on VSTS. In VSTS I'm using the Azure CLI task and I have to run "call az login -u $(DeployUsername) -p $(DeployPassword)" rather than letting the Azure CLI task log in using a service principle.

**Comment by tombuildsstuff** [comment]
👋 hey folks!

So taking another look at this, I think there's two separate issues here:

1. authenticating as a Service Principal via the Azure CLI

At this time we only support authenticating as Service Principals when using the methods outlined in the ["Authenticating using a Service Principal" page](https://www.terraform.io/docs/providers/azurerm/authenticating_via_service_principal.html). Authenticating using a Service Principal and the Azure CLI is not supported (and support isn't planned at this time).

We've recently updated our documentation to state this explicitly - but I've opened #1116 which clarifies this further.

2. Authenticating using the Azure CLI and specifying the Tenant (e.g. `az login --tenant foo.onmicrosoft.com`)

This isn't currently supported, but should be - and is being tracked in #1029, as such I'm going to close this issue in favour of that one.

Thanks!
